### PR TITLE
Check for datadir after the config files were read

### PR DESCRIPTION
**Pull request by Flowdalic**
If we first check for datadir existence and then read the config file
we bail out in situations where the default datadir is invalid,
e.g. because of an non-existing HOME, but the datadir specified in
config files is valid. We should not do that, as the user may specified a
valid datadir in the config file and expects the setting to take effect.

See for example https://github.com/bitcoin/bitcoin/pull/12255#issuecomment-403666092 where users hit this.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Doesn't the location of the default config file depend on the datadir?

**Comment by Flowdalic** [comment]
> Doesn't the location of the default config file depend on the datadir?

Does not appear to be the case: 

```
BITCOIN_CONF_FILENAME = "bitcoin.conf"
```

is used if not explicitly specified via `-conf`. Which would mean that it is opened relative to the working directory of the process. But even if the location of the default config file is dependent on the datadir, I don't see a problem.  If I'm not overlooking something.

**Comment by MarcoFalke** [comment]
See https://dev.visucore.com/bitcoin/doxygen/class_args_manager.html#aec848fd54ace0df3df48dac446feb7d2 which calls `GetDataDir`

**Comment by AtsukiTak** [comment]
Now the second if clause seems meaningless.

https://github.com/Flowdalic/bitcoin/blob/94a038b8694229c0153ede48a53701424ca51e38/src/bitcoind.cpp#L98-L102

Because it already checked in `ReadConfigFiles`.

https://github.com/bitcoin/bitcoin/blob/4a7e64fc85461a205f2b51da52d1455795d43b91/src/util.cpp#L909-L912

**Comment by Flowdalic** [comment]
> Now the second if clause seems meaningless.

But it isn't: The datadir may have been set in a configuration file, hence we need to check after the config files have been read.

**Comment by Flowdalic** [comment]
Right, the check can get probably removed.

**Comment by Flowdalic** [comment]
Rebased. Also fixed the cause and re-enabled the tests which were disabled in https://github.com/bitcoin/bitcoin/pull/13687 with fabe28a0cdcfa13e0e595a0905e3642a960d3077.

I had a hard time working on the argument and configuration file parsing code. It is a little bit spagettish and some parts are messy.

I used default argument values in order to optionally change GetDataDir() to not create the data directory, which, I assumed, made the change less intrusive. But I am happy to change it towards a GetDataDir() / GetDataDirAndCreateIfNecessary(). I am undecied which of those two approaches would be better.

**Comment by Flowdalic** [comment]
The failed test `wallet_import_rescan.py` is green here. Could some restart anther travis run?

File: src/bitcoin-cli.cpp [hunk 1]
```diff
@@ -121,14 +121,14 @@ static int AppInitRPC(int argc, char* argv[])
...
     }
-    if (!fs::is_directory(GetDataDir(false))) {
-        fprintf(stderr, "Error: Specified data directory \"%s\" does not exist.\n", gArgs.GetArg("-datadir", "").c_str());
-        return EXIT_FAILURE;
-    }
     if (!gArgs.ReadConfigFiles(error, true)) {
         fprintf(stderr, "Error reading configuration file: %s\n", error.c_str());
         return EXIT_FAILURE;
     }
+    if (!fs::is_directory(GetDataDir(false))) {
```
Comment (AtsukiTak) [inline src/bitcoin-cli.cpp]: It seems never fail unless user specify unpermitted path.

Comment (Flowdalic) [reply src/bitcoin-cli.cpp]: It fails if the specified path is not a directory. I think that is sensible and was already previous behavior. Or do have a specific change request?

**Comment by Flowdalic** [comment]
> I think you should leave only first one

Well, those commits depend on each other: For example I wouldn't be able to re-enable the relevant tests of "Check for datadir after the config files were read" without the previous commits.

I don't feel like that this patch series is worth splitting up, as it is pretty small, but I am happy create single PRs for each commit one after another, if there is a consensus that this would be favorable.

File: src/bitcoin-cli.cpp (hunk 1 above)
Comment (AtsukiTak) [reply src/bitcoin-cli.cpp]: If I specify a file, `boost::filesystem::create_directory` throw an error.
If I specify a non-exist path, a new directory is created and no error happens.

I think you just missed second argument; `GetDataDir(false, false)` might be what you want.

Comment (AtsukiTak) [reply src/bitcoin-cli.cpp]: Sorry for my poor English. Please feel free to ask me if my English is not clear.

**Comment by AtsukiTak** [comment]
> I don't feel like that this patch series is worth splitting up, as it is pretty small

I see. So please don't care about that. That was just newbie's nonsense comment.

**Comment by leishman** [comment]
I don't think our PRs conflict, but just wanted to make you aware of a minor change I'm putting in the logging behavior: https://github.com/bitcoin/bitcoin/pull/14057

I'm also working on logic to create a `bitcoin.conf.example` file: https://github.com/bitcoin/bitcoin/pull/13761

**Comment by Flowdalic** [comment]
Pinging @MarcoFalke since this PR re-enables tests he disabled in fabe28a0cdcfa13e0e595a0905e3642a960d3077. Would be great to hear your thoughts on this.

**Comment by MarcoFalke** [comment]
I haven't looked at the most recent code, but I still think that the location of the default config file depends on the datadir? See https://dev.visucore.com/bitcoin/doxygen/class_args_manager.html#aec848fd54ace0df3df48dac446feb7d2 which calls `GetDataDir`.

So you are changing how and what config files are read. Not sure if we want this behavior change without discussion first.

**Comment by MarcoFalke** [comment]
Also, you modified how bitcoind and bitcoin-cli work, but not the gui. Wouldn't that lead to issues when the gui is used as rpc server via the bitcoin-cli?

**Comment by fanquake** [comment]
Closing "up for grabs".

**Pull request closed without merging**